nping/docs/nping.1
'\" t
.\"     Title: nping
.\"    Author: [see the "Authors" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03/27/2026
.\"    Manual: Nping Reference Guide
.\"    Source: Nping
.\"  Language: English
.\"
.TH "NPING" "1" "03/27/2026" "Nping" "Nping Reference Guide"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
nping \- Network packet generation tool / ping utility
.SH "SYNOPSIS"
.HP \w'\fBnping\fR\ 'u
\fBnping\fR [\fIOptions\fR] {\fItargets\fR}
.SH "DESCRIPTION"
.PP
Nping is an open\-source tool for network packet generation, response analysis and response time measurement\&. Nping allows users to generate network packets of a wide range of protocols, letting them tune virtually any field of the protocol headers\&. While Nping can be used as a simple ping utility to detect active hosts, it can also be used as a raw packet generator for network stack stress tests, ARP poisoning, Denial of Service attacks, route tracing, and other purposes\&.
.PP
Additionally, Nping offers a special mode of operation called the "Echo Mode", that lets users see how the generated probes change in transit, revealing the differences between the transmitted packets and the packets received at the other end\&. See section "Echo Mode" for details\&.
.PP
The output from Nping is a list of the packets that are being sent and received\&. The level of detail depends on the options used\&.
.PP
A typical Nping execution is shown in
Example\ \&1\&. The only Nping arguments used in this example are
\fB\-c\fR, to specify the number of times to target each host,
\fB\-\-tcp\fR
to specify TCP Probe Mode,
\fB\-p 80,433\fR
to specify the target ports; and then the two target hostnames\&.
.PP
\fBExample\ \&1.\ \&A representative Nping execution\fR
.sp
.if n \{\
.RS 4
.\}
.nf
# \fBnping \-c 1 \-\-tcp \-p 80,433 scanme\&.kmap\&.org google\&.com\fR

Starting Nping ( https://kmap\&.org/nping )
SENT (0\&.0120s) TCP 96\&.16\&.226\&.135:50091 > 64\&.13\&.134\&.52:80 S ttl=64 id=52072 iplen=40  seq=1077657388 win=1480 
RCVD (0\&.1810s) TCP 64\&.13\&.134\&.52:80 > 96\&.16\&.226\&.135:50091 SA ttl=53 id=0 iplen=44  seq=4158134847 win=5840 <mss 1460>
SENT (1\&.0140s) TCP 96\&.16\&.226\&.135:50091 > 74\&.125\&.45\&.100:80 S ttl=64 id=13932 iplen=40  seq=1077657388 win=1480 
RCVD (1\&.1370s) TCP 74\&.125\&.45\&.100:80 > 96\&.16\&.226\&.135:50091 SA ttl=52 id=52913 iplen=44  seq=2650443864 win=5720 <mss 1430>
SENT (2\&.0140s) TCP 96\&.16\&.226\&.135:50091 > 64\&.13\&.134\&.52:433 S ttl=64 id=8373 iplen=40  seq=1077657388 win=1480 
SENT (3\&.0140s) TCP 96\&.16\&.226\&.135:50091 > 74\&.125\&.45\&.100:433 S ttl=64 id=23624 iplen=40  seq=1077657388 win=1480 

Statistics for host scanme\&.kmap\&.org (64\&.13\&.134\&.52):
 |  Probes Sent: 2 | Rcvd: 1 | Lost: 1  (50\&.00%)
 |_ Max rtt: 169\&.720ms | Min rtt: 169\&.720ms | Avg rtt: 169\&.720ms
Statistics for host google\&.com (74\&.125\&.45\&.100):
 |  Probes Sent: 2 | Rcvd: 1 | Lost: 1  (50\&.00%)
 |_ Max rtt: 122\&.686ms | Min rtt: 122\&.686ms | Avg rtt: 122\&.686ms
Raw packets sent: 4 (160B) | Rcvd: 2 (92B) | Lost: 2 (50\&.00%)
Tx time: 3\&.00296s | Tx bytes/s: 53\&.28 | Tx pkts/s: 1\&.33
Rx time: 3\&.00296s | Rx bytes/s: 30\&.64 | Rx pkts/s: 0\&.67
Nping done: 2 IP addresses pinged in 4\&.01 seconds
.fi
.if n \{\
.RE
.\}
.PP
The newest version of Nping can be obtained with Kmap at
\m[blue]\fB\%https://github.com/YurilLAB/Kmap\fR\m[]\&. The newest version of this man page is available at
\m[blue]\fB\%https://github.com/YurilLAB/Kmap/book/nping-man.html\fR\m[]\&.
.SH "OPTIONS SUMMARY"
.PP
This options summary is printed when Nping is run with no arguments\&. It helps people remember the most common options, but is no substitute for the in\-depth documentation in the rest of this manual\&. Some obscure options aren\*(Aqt even included here\&.
.sp
.if n \{\
.RS 4
.\}
.nf
Nping 0\&.7\&.92SVN ( https://kmap\&.org/nping )
Usage: nping [Probe mode] [Options] {target specification}

TARGET SPECIFICATION:
  Targets may be specified as hostnames, IP addresses, networks, etc\&.
  Ex: scanme\&.kmap\&.org, microsoft\&.com/24, 192\&.168\&.0\&.1; 10\&.0\&.*\&.1\-24
PROBE MODES:
  \-\-tcp\-connect                    : Unprivileged TCP connect probe mode\&.
  \-\-tcp                            : TCP probe mode\&.
  \-\-udp                            : UDP probe mode\&.
  \-\-icmp                           : ICMP probe mode\&.
  \-\-arp                            : ARP/RARP probe mode\&.
  \-\-tr, \-\-traceroute               : Traceroute mode (can only be used with 
                                     TCP/UDP/ICMP modes)\&.
TCP CONNECT MODE:
   \-p, \-\-dest\-port <port spec>     : Set destination port(s)\&.
   \-g, \-\-source\-port <portnumber>  : Try to use a custom source port\&.
TCP PROBE MODE:
   \-g, \-\-source\-port <portnumber>  : Set source port\&.
   \-p, \-\-dest\-port <port spec>     : Set destination port(s)\&.
   \-\-seq <seqnumber>               : Set sequence number\&.
   \-\-flags <flag list>             : Set TCP flags (ACK,PSH,RST,SYN,FIN\&.\&.\&.)
   \-\-ack <acknumber>               : Set ACK number\&.
   \-\-win <size>                    : Set window size\&.
   \-\-badsum                        : Use a random invalid checksum\&. 
UDP PROBE MODE:
   \-g, \-\-source\-port <portnumber>  : Set source port\&.
   \-p, \-\-dest\-port <port spec>     : Set destination port(s)\&.
   \-\-badsum                        : Use a random invalid checksum\&. 
ICMP PROBE MODE:
  \-\-icmp\-type <type>               : ICMP type\&.
  \-\-icmp\-code <code>               : ICMP code\&.
  \-\-icmp\-id <id>                   : Set identifier\&.
  \-\-icmp\-seq <n>                   : Set sequence number\&.
  \-\-icmp\-redirect\-addr <addr>      : Set redirect address\&.
  \-\-icmp\-param\-pointer <pnt>       : Set parameter problem pointer\&.
  \-\-icmp\-advert\-lifetime <time>    : Set router advertisement lifetime\&.
  \-\-icmp\-advert\-entry <IP,pref>    : Add router advertisement entry\&.
  \-\-icmp\-orig\-time  <timestamp>    : Set originate timestamp\&.
  \-\-icmp\-recv\-time  <timestamp>    : Set receive timestamp\&.
  \-\-icmp\-trans\-time <timestamp>    : Set transmit timestamp\&.
ARP/RARP PROBE MODE:
  \-\-arp\-type <type>                : Type: ARP, ARP\-reply, RARP, RARP\-reply\&.
  \-\-arp\-sender\-mac <mac>           : Set sender MAC address\&.
  \-\-arp\-sender\-ip  <addr>          : Set sender IP address\&.
  \-\-arp\-target\-mac <mac>           : Set target MAC address\&.
  \-\-arp\-target\-ip  <addr>          : Set target IP address\&.
IPv4 OPTIONS:
  \-S, \-\-source\-ip                  : Set source IP address\&.
  \-\-dest\-ip <addr>                 : Set destination IP address (used as an 
                                     alternative to {target specification} )\&. 
  \-\-tos <tos>                      : Set type of service field (8bits)\&.
  \-\-id  <id>                       : Set identification field (16 bits)\&.
  \-\-df                             : Set Don\*(Aqt Fragment flag\&.
  \-\-mf                             : Set More Fragments flag\&.
  \-\-evil                           : Set Reserved / Evil flag\&.
  \-\-ttl <hops>                     : Set time to live [0\-255]\&.
  \-\-badsum\-ip                      : Use a random invalid checksum\&. 
  \-\-ip\-options <S|R [route]|L [route]|T|U \&.\&.\&.> : Set IP options
  \-\-ip\-options <hex string>                    : Set IP options
  \-\-mtu <size>                     : Set MTU\&. Packets get fragmented if MTU is
                                     small enough\&.
IPv6 OPTIONS:
  \-6, \-\-IPv6                       : Use IP version 6\&.
  \-\-dest\-ip                        : Set destination IP address (used as an
                                     alternative to {target specification})\&.
  \-\-hop\-limit                      : Set hop limit (same as IPv4 TTL)\&.
  \-\-traffic\-class <class> :        : Set traffic class\&.
  \-\-flow <label>                   : Set flow label\&.
ETHERNET OPTIONS:
  \-\-dest\-mac <mac>                 : Set destination mac address\&. (Disables
                                     ARP resolution)
  \-\-source\-mac <mac>               : Set source MAC address\&.
  \-\-ether\-type <type>              : Set EtherType value\&.
PAYLOAD OPTIONS:
  \-\-data <hex string>              : Include a custom payload\&.
  \-\-data\-string <text>             : Include a custom ASCII text\&.
  \-\-data\-length <len>              : Include len random bytes as payload\&.
ECHO CLIENT/SERVER:
  \-\-echo\-client <passphrase>       : Run Nping in client mode\&.
  \-\-echo\-server <passphrase>       : Run Nping in server mode\&.
  \-\-echo\-port <port>               : Use custom <port> to listen or connect\&.
  \-\-no\-crypto                      : Disable encryption and authentication\&.
  \-\-once                           : Stop the server after one connection\&.
  \-\-safe\-payloads                  : Erase application data in echoed packets\&.
TIMING AND PERFORMANCE:
  Options which take <time> are in seconds, or append \*(Aqms\*(Aq (milliseconds),
  \*(Aqs\*(Aq (seconds), \*(Aqm\*(Aq (minutes), or \*(Aqh\*(Aq (hours) to the value (e\&.g\&. 30m, 0\&.25h)\&.
  \-\-delay <time>                   : Adjust delay between probes\&.
  \-\-rate  <rate>                   : Send num packets per second\&.
MISC:
  \-h, \-\-help                       : Display help information\&.
  \-V, \-\-version                    : Display current version number\&. 
  \-c, \-\-count <n>                  : Stop after <n> rounds\&.
  \-e, \-\-interface <name>           : Use supplied network interface\&.
  \-H, \-\-hide\-sent                  : Do not display sent packets\&.
  \-N, \-\-no\-capture                 : Do not try to capture replies\&.
  \-\-privileged                     : Assume user is fully privileged\&.
  \-\-unprivileged                   : Assume user lacks raw socket privileges\&.
  \-\-send\-eth                       : Send packets at the raw Ethernet layer\&.
  \-\-send\-ip                        : Send packets using raw IP sockets\&.
  \-\-bpf\-filter <filter spec>       : Specify custom BPF filter\&.
OUTPUT:
  \-v                               : Increment verbosity level by one\&.
  \-v[level]                        : Set verbosity level\&. E\&.g: \-v4
  \-d                               : Increment debugging level by one\&.
  \-d[level]                        : Set debugging level\&. E\&.g: \-d3
  \-q                               : Decrease verbosity level by one\&.
  \-q[N]                            : Decrease verbosity level N times
  \-\-quiet                          : Set verbosity and debug level to minimum\&.
  \-\-debug                          : Set verbosity and debug to the max level\&.
EXAMPLES:
  nping scanme\&.kmap\&.org
  nping \-\-tcp \-p 80 \-\-flags rst \-\-ttl 2 192\&.168\&.1\&.1
  nping \-\-icmp \-\-icmp\-type time \-\-delay 500ms 192\&.168\&.254\&.254
  nping \-\-echo\-server "public" \-e wlan0 \-vvv 
  nping \-\-echo\-client "public" echo\&.kmap\&.org \-\-tcp \-p1\-1024 \-\-flags ack

SEE THE MAN PAGE FOR MANY MORE OPTIONS, DESCRIPTIONS, AND EXAMPLES

.fi
.if n \{\
.RE
.\}
.SH "TARGET SPECIFICATION"
.PP
Everything on the Nping command line that isn\*(Aqt an option or an option argument is treated as a target host specification\&. Nping uses the same syntax for target specifications that Kmap does\&. The simplest case is a single target given by IP address or hostname\&.
.PP
Nping supports CIDR\-style
addressing\&. You can append
/\fInumbits\fR
to an IPv4 address or hostname and Nping will send probes to every IP address for which the first
\fInumbits\fR
are the same as for the reference IP or hostname given\&. For example,
192\&.168\&.10\&.0/24
would send probes to the 256 hosts between 192\&.168\&.10\&.0 (binary:
11000000 10101000 00001010 00000000) and 192\&.168\&.10\&.255 (binary:
11000000 10101000 00001010 11111111), inclusive\&.
192\&.168\&.10\&.40/24
would ping exactly the same targets\&. Given that the host scanme\&.kmap\&.org
is at the IP address 64\&.13\&.134\&.52, the specification
scanme\&.kmap\&.org/16
would send probes to the 65,536 IP addresses between 64\&.13\&.0\&.0 and 64\&.13\&.255\&.255\&. The smallest allowed value is
/0, which targets the whole Internet\&. The largest value is
/32, which targets just the named host or IP address because all address bits are fixed\&.
.PP
CIDR notation is short but not always flexible enough\&. For example, you might want to send probes to 192\&.168\&.0\&.0/16 but skip any IPs ending with \&.0 or \&.255 because they may be used as subnet network and broadcast addresses\&. Nping supports this through octet range addressing\&. Rather than specify a normal IP address, you can specify a comma\-separated list of numbers or ranges for each octet\&. For example,
192\&.168\&.0\-255\&.1\-254
will skip all addresses in the range that end in \&.0 or \&.255, and
192\&.168\&.3\-5,7\&.1
will target the four addresses 192\&.168\&.3\&.1, 192\&.168\&.4\&.1, 192\&.168\&.5\&.1, and 192\&.168\&.7\&.1\&. Either side of a range may be omitted; the default values are 0 on the left and 255 on the right\&. Using
\-
by itself is the same as
0\-255, but remember to use
0\-
in the first octet so the target specification doesn\*(Aqt look like a command\-line option\&. Ranges need not be limited to the final octets: the specifier
0\-\&.\-\&.13\&.37
will send probes to all IP addresses on the Internet ending in \&.13\&.37\&. This sort of broad sampling can be useful for Internet surveys and research\&.
.PP
IPv6 addresses can only be specified by their fully qualified IPv6 address or hostname\&. CIDR and octet ranges aren\*(Aqt supported for IPv6 because they are rarely useful\&.
.PP
Nping accepts multiple host specifications on the command line, and they don\*(Aqt need to be the same type\&. The command
\fBnping scanme\&.kmap\&.org 192\&.168\&.0\&.0/8 10\&.0\&.0,1,3\-7\&.\-\fR
does what you would expect\&.
.SH "OPTION SPECIFICATION"
.PP
Nping is designed to be very flexible and fit a wide variety of needs\&. As with most command\-line tools, its behavior can be adjusted using command\-line options\&. These general principles apply to option arguments, unless stated otherwise\&.
.PP
Options that take integer numbers can accept values specified in decimal, octal or hexadecimal base\&. When a number starts with
0x, it will be treated as hexadecimal; when it simply starts with
0, it will be treated as octal\&. Otherwise, Nping will assume the number has been specified in base 10\&. Virtually all numbers that can be supplied from the command line are unsigned so, as a general rule, the minimum value is zero\&. Users may also specify the word
random
or
rand
to make Nping generate a random value within the expected range\&.
.PP
IP addresses may be given as IPv4 addresses (e\&.g\&.
192\&.168\&.1\&.1), IPv6 addresses (e\&.g\&.
2001:db8:85a3::8e4c:760:7146), or hostnames, which will be resolved using the default DNS server configured in the host system\&.
.PP
Options that take MAC addresses accept the usual colon\-separated 6 hex byte format (e\&.g\&.
00:50:56:d4:01:98)\&. Hyphens may also be used instead of colons (e\&.g\&.
00\-50\-56\-c0\-00\-08)\&. The special word
random
or
rand
sets a random address and the word
broadcast
or
bcast
sets ff:ff:ff:ff:ff:ff\&.
.SH "GENERAL OPERATION"
.PP
Unlike other ping and packet generation tools, Nping supports multiple target host and port specifications\&. While this provides great flexibility, it is not obvious how Nping handles situations where there is more than one host and/or more than one port to send probes to\&. This section explains how Nping behaves in these cases\&.
.PP
When multiple target hosts are specified, Nping rotates among them in round\-robin fashion\&. This gives slow hosts more time to send their responses before another probe is sent to them\&. Ports are also scheduled using round robin\&. So, unless only one port is specified, Nping never sends two probes to the same target host and port consecutively\&.
.PP
The loop around targets is the
\(lqinner loop\(rq
and the loop around ports is the

CHANGELOG
#Kmap Changelog ($Id$); -*-text-*-

o Fix a 1-byte overrun (read) while reading certain crafted DNS labels,
  reported by Peter Parker. [Daniel Miller]

o [NSE][GH#3317] Function url.build_path was mangling special characters in URL
  path segments. [nnposter]

Kmap 7.99 [2026-03-26]

o Integrated many of the most-frequently-submitted IPv4 and IPv6 OS
  fingerprints, as well as dozens of updated service fingerprints.

o Upgraded included libraries: OpenSSL 3.0.19, libpcap 1.10.6, libpcre2 10.47,
  liblinear 2.50, zlib 1.3.2

o [Windows] Upgraded the included version of Npcap from 1.83 to 1.87, resolving
  several crashes and stability-related issues. See https://npcap.com/changelog

o [Zekmap][GH#3182] Zekmap is now distributed as a universal wheel
  (zekmap-7.99-py3-none-any.whl) instead of an RPM package so that it can be
  installed on any system with Python 3. [Daniel Miller]

o [Ncat][Windows] Limited the number of handles inherited by subprocesses
  launched with -e, preventing interference between clients when -e and
  --keep-open are used. Reported by Nimish Verma.

o [Ncat] Several fixes for regressions or longstanding failure cases in
  ncat-test.pl [Daniel Miller]:

  + [Windows] Fixed handling of socket EOF with --exec

  + Fixed the -i (idle timeout) option for listen mode, which was broken
    when adding the -q option in Ncat 7.96

  + Fixed HTTP proxy server when SSL is used.

  + DTLS (SSL over UDP) shutdown connection on stdin EOF.

o [Windows][GH#2711] Kmap now supports scanning over various VPN virtual
  adapters like OpenVPN TAP adapters. [Daniel Miller]

o [GH#3280] Fix a performance regression in reverse-DNS in Kmap 7.98. The fix
  for #3130 had caused Kmap to send requests too slowly. [Daniel Miller]

o [macOS][GH#3289] Fixed a configure-time failure in libdnet that resulted in
  incorrect MAC addresses being reported. [Daniel Miller]

o [Zekmap][GH#3189] Fix a crash in Zekmap topology and hosts viewer:
  "TypeError: format requires a mapping" [Daniel Miller]

o [GH#2955] Fix a routing issue with -e and -S related to #2206
  causing error "setup_target: failed to determine route" [Daniel Miller]

o [GH#3214] Improve compatibility of build process on various platforms and add
  multiplatform autobuilds in Github workflow. [Jordan Ritter]

o [NSE][GH#2183][GH#3239] Script hostmap-crtsh now reports only true subdomains
  of a given target hostname by default. In the past, it was reporting any
  DNS name that included the target hostname as a substring (but not
  necessarily as a suffix). The old behavior can be enabled by setting script
  argument hostmap-crtsh.lax. [Sweekar-cmd, nnposter]

o [NSE] Function url.parse_query was not interpreting plus signs as spaces.
  [nnposter]

o [NSE] Function url.parse was not properly parsing URLs with query strings
  but empty paths. [nnposter]

o [NSE][GH#3287] Functions tableaux.tcopy and tableaux.shallow_tcopy were
  not behaving the same when the input table had a custom __pairs metamethod.
  Both functions now perform a raw copy, ignoring the metamethod. [nnposter]

o [NSE] Function tableaux.shallow_tcopy did not work correctly for tables
  with Boolean keys. [nnposter]

o [NSE] IPP print queue job details were not getting populated, having
  a hard dependency on Apple-specific attributes. [nnposter]

o [NSE][GH#3245] Functions connect and close have been removed from the IPP
  library, as they served no purpose. [nnposter]

o [NSE] ipOps.expand_ip was crashing upon malformed IPv6 addresses. [nnposter]

o [NSE][GH#3262] FTP banner parsing is now more closely aligned with RFC 959,
  section 4.2. [nnposter]

o [NSE][GH#3253] Function stdnse.make_buffer now accepts an extra parameter
  that allows preloading the newly created buffer with data. [nnposter]

o [NSE][GH#3191][GH#3218] Script http-internal-ip-disclosure has been enhanced,
  including added support for IPv6 and HTTPS and more accurate processing
  of target responses. [nnposter]

o [NSE][GH#3194] RPC-based scripts were sporadically failing due to privileged
  port conflicts. [nnposter]

o [NSE][GH#3196] Script rlogin-brute was sporadically failing due to using
  an off-by-one range for privileged ports and not handling potential
  port conflicts. [nnposter]

Kmap 7.98 [2025-08-21]

o [SECURITY] Rebuilt the Windows self-installer with NSIS 3.11, addressing
  CVE-2025-43715--a race condition in earlier NSIS versions that could allow
  local attackers to escalate to SYSTEM privileges when a vulnerable installer is
  run as SYSTEM. The Kmap installer does not run as SYSTEM by default.

o Upgraded included libraries: OpenSSL 3.0.17, Lua 5.4.8

o [Windows] Upgraded the included version of Npcap from 1.82 to 1.83, improving
  compatibility with PPPoE connections. See https://npcap.com/changelog

o [macOS][GH#3127] Fix "dnet: Failed to open device en0" errors on macOS since
  Kmap 7.96. [Daniel Miller]

o Fixed an issue in FTP bounce scan where a single null byte is written past
  the end of the receive buffer. The issue is triggered by a malicious server
  but does not cause a crash with default builds. [Tyler Zars]

o [GH#3130] Fix a crash (stack exhaustion due to excessive recursion) in the
  parallel DNS resolver. Additionally, improved performance by processing
  responses that come after the request has timed out. [Daniel Miller]

o [GH#2148] Fix the error, "Assertion failed: (datalink == DLT_EN10MB), function begin_sniffer, file scan_engine_raw.cc"
  when using Kmap with certain VPN interfaces. [Daniel Miller]

o [GH#2757] Fix a crash in traceroute when using randomly-generated decoys:
  "Assertion `source->ss_family == AF_INET' failed" [Daniel Miller]

o [GH#2899] When IP protocol scanning on IPv6 (-sO -6), skip protocol numbers
  that are registered as Extension Header values. When the --data option was
  used, these would fail the assertion "len == (u32) ntohs(ip6->ip6_plen)"
  [Daniel Miller]


nping/nping.cc

/***************************************************************************
 * Nping.cc -- This file contains function main() and some other general   *
 * high level functions.                                                   *
 *                                                                         *
 ***********************IMPORTANT KMAP LICENSE TERMS************************
 *
 * The Kmap Security Scanner is (C) 1996-2026 Kmap Software LLC ("The Kmap
 * Project"). Kmap is also a registered trademark of the Kmap Project.
 *
 * This program is distributed under the terms of the Kmap Public Source
 * License (NPSL). The exact license text applying to a particular Kmap
 * release or source code control revision is contained in the LICENSE
 * file distributed with that version of Kmap or source code control
 * revision. More Kmap copyright/legal information is available from
 * https://github.com/YurilLAB/Kmap/blob/master/LICENSE, and further information on the
 * NPSL license itself can be found at https://github.com/YurilLAB/Kmap/blob/master/LICENSE . This
 * header summarizes some key points from the Kmap license, but is no
 * substitute for the actual license text.
 *
 * Kmap is generally free for end users to download and use themselves,
 * including commercial use. It is available from https://github.com/YurilLAB/Kmap.
 *
 * The Kmap license generally prohibits companies from using and
 * redistributing Kmap in commercial products, but we sell a special Kmap
 * OEM Edition with a more permissive license and special features for
 * this purpose. See https://github.com/YurilLAB/Kmap
 *
 * If you have received a written Kmap license agreement or contract
 * stating terms other than these (such as an Kmap OEM license), you may
 * choose to use and redistribute Kmap under those terms instead.
 *
 * The official Kmap Windows builds include the Npcap software
 * (https://npcap.com) for packet capture and transmission. It is under
 * separate license terms which forbid redistribution without special
 * permission. So the official Kmap Windows builds may not be redistributed
 * without special permission (such as an Kmap OEM license).
 *
 * Source is provided to this software because we believe users have a
 * right to know exactly what a program is going to do before they run it.
 * This also allows you to audit the software for security holes.
 *
 * Source code also allows you to port Kmap to new platforms, fix bugs, and
 * add new features. You are highly encouraged to submit your changes as a
 * Github PR at https://github.com/YurilLAB/Kmap for possible incorporation
 * into the main distribution. Unless you specify otherwise, it
 * is understood that you are offering us very broad rights to use your
 * submissions as described in the Kmap Public Source License Contributor
 * Agreement. This is important because we fund the project by selling licenses
 * with various terms, and also because the inability to relicense code has
 * caused devastating problems for other Free Software projects (such as KDE
 * and NASM).
 *
 * The free version of Kmap is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. Warranties,
 * indemnification and commercial support are all available through the
 * Kmap project -- see https://github.com/YurilLAB/Kmap
 *
 ***************************************************************************/

#ifdef WIN32
#include "winfix.h"
#endif

#include "nping.h"
#include "output.h"
#include "NpingOps.h"
#include "utils.h"
#include "utils_net.h"
#include "nsock.h"
#include "global_structures.h"
#include "ArgParser.h"
#include "EchoHeader.h"
#include "EchoClient.h"
#include "EchoServer.h"
#include "ProbeMode.h"
#include "common.h"
#include "dnet.h"
#include "pcap.h"
#include <signal.h>
#include <time.h>

NpingOps o;
EchoClient ec;
EchoServer es;

int do_safe_checks();
void test_stuff();
void signal_handler(int signo);


/** Main function. It basically inits Nping working environment, calls the
  * command-line argument parser, and enters the appropriate mode (normal
  * probe mode, echo client or echo server). */
int main(int argc, char *argv[] ){

  struct tm tm;    /* For time display                */
  int err;
  time_t now;       /* Stores current time             */
  char tbuf[128];   /* Stores current time as a string */
  ArgParser a;      /* Command line argument parser    */
  unsigned long int i=0;
  ProbeMode prob;
  NpingTarget *t=NULL;

  /* Get current time */
  tzset();
  now = time(NULL);
  err = n_localtime(&now, &tm);
  if (err)
    nping_fatal(QT_3,"Error in localtime: %s", strerror(err));
  o.stats.startRuntime();

  /* Some run-time tests to ensure everything works as expected */
  do_safe_checks();

  /* Init a few things on Windows */
  #ifdef WIN32
    win_pre_init();
  #endif

  /* Register the SIGINT signal so when the users presses CTRL-C we print stats
   * before quitting. */
  #if HAVE_SIGNAL
    signal(SIGINT, signal_handler); 
  #endif

  /* Let's parse and validate user supplied args */
  a.parseArguments(argc, argv);
  #ifdef WIN32
    // Must come after parseArguments because of --unprivileged
    // Must come before validateOptions because it sets o.isRoot
    win_init();
  #endif
  o.validateOptions();

  /* ISO 8601 date/time -- http://www.cl.cam.ac.uk/~mgk25/iso-time.html */
  if ( strftime(tbuf, sizeof(tbuf), "%Y-%m-%d %H:%M %Z", &tm) <= 0)
    nping_fatal(QT_3,"Unable to properly format time");
  nping_print(QT_1, "\nStarting %s %s ( %s ) at %s", NPING_NAME, NPING_VERSION, NPING_URL, tbuf);

  /*If nping is called on something that doesn't take port scanning
   * we should alert the user that their port command is going to be ignored
   * I choose to print out a Fatal error since the scan doesn't make sense.
   */
  if(o.issetTargetPorts() && !o.scan_mode_uses_target_ports(o.getMode()))
      nping_fatal(QT_3, "You cannot use -p (explicit port selection) in your current scan mode.\n(Perhaps you meant to use --tcp or --udp)");



  /* Resolve and cache target specs */
  nping_print(DBG_2,"Resolving specified targets...");
  o.targets.processSpecs();
  if( ((i=o.targets.getTargetsFetched())<=0) && o.getRole()!=ROLE_SERVER )
    nping_fatal(QT_3, "Execution aborted. Nping needs at least one valid target to operate.");
  else
    nping_print(DBG_2,"%lu target IP address%s determined.", i, (i==1)? "":"es" );

  switch( o.getRole() ){

        case ROLE_NORMAL:
            prob.start();
            prob.cleanup();
        break;

        case ROLE_CLIENT:
            t=o.targets.getNextTarget();
            o.targets.rewind();
            ec.start(t, o.getEchoPort() );
            ec.cleanup();
        break;

        case ROLE_SERVER:
            o.stats.startClocks();
            es.start();
            // Cleanup currently does nothing, but needs to be called in case
            // it does something in the future.
            es.cleanup(); // lgtm [cpp/useless-expression]
            o.stats.stopClocks();
        break;

        default:
            nping_fatal(QT_3, "Invalid role %d\n", o.getRole() );
        break;
  }

  /* Display stats, clean up and quit */ 
  o.stats.stopRuntime();
  o.displayStatistics();
  o.displayNpingDoneMsg();
  o.cleanup();
  exit(EXIT_SUCCESS);

  exit(EXIT_SUCCESS);
} /* End of main() */


/* Things that should be guaranteed by the compiler, standard library, OS etc,
 *  but that we check just in case... */
int do_safe_checks(){
 if( (sizeof(u32) != 4) || (sizeof(u16) != 2) || (sizeof(u8) != 1) )
    nping_fatal(QT_3,"Types u32, u16 and u8 do not have the correct sizes on your system.");
  test_stuff(); /* Little function that is called quite early to test some misc stuff. */
  return OP_SUCCESS;
} /* End of do_safe_checks() */



/** Use this function whenever you have some code that you want to test, but
  * you don't want to bother creating a new dummy main.c file, etc. This
  * function is called by do_safe_checks() at the beginning of nping execution.
  * Command line arguments have not been parsed yet so even if you run nping
  * with no arguments you will not get an error, and this function will be
  * called. You probably want to place an exit() call at the end of your
  * testing code so Nping does not actually continue its normal execution path
  * but exit after your tests.  */
void test_stuff(){
  return;
} /* End of test_stuff() */


/** This function is called whenever user presses CTRL-C. Basically what we
  * do here is stop Tx and Rx clocks, stop global clock, display statistics,
  * do a bit of cleanup and exit the program. The exit() call makes the
  * program return EXIT_FAILURE instead of the usual EXIT_SUCCESS.
  *
  * TODO: Many of the things done in this function may not be safe due to
  * reentrancy issues. Check http://seclists.org/kmap-dev/2009/q3/0596.html
  * and http://seclists.org/kmap-dev/2009/q3/0596.html */
void signal_handler(int signo){
  fflush(stdout);
  nping_print(DBG_1,"signal_handler(): Received signal %d", signo);
  switch(signo) {
      case SIGINT:
        o.stats.stopTxClock();
        o.stats.stopRxClock();
        o.stats.stopRuntime();
        o.displayStatistics();
        o.displayNpingDoneMsg();
        o.cleanup();
        fflush(stderr);
        exit(EXIT_FAILURE);
      break;

      default:
        nping_warning(QT_2, "signal_handler(): Unexpected signal received (%d). Please report a bug.", signo);
      break;
  }
  fflush(stderr);
  exit(EXIT_FAILURE);
} /* End of signal_handler() */
